Sources/SwiftParser/Patterns.swift

```diff
@@ -120,20 +120,31 @@ extension Parser {
   /// =======
   ///
   ///     typed-pattern → pattern ':' attributes? inout? type
-  mutating func parseTypedPattern() -> (RawPatternSyntax, RawTypeAnnotationSyntax?) {
+  mutating func parseTypedPattern(allowRecover: Bool = true) -> (RawPatternSyntax, RawTypeAnnotationSyntax?) {
     let pattern = self.parsePattern()
 
     // Now parse an optional type annotation.
-    guard let colon = self.consume(if: .colon) else {
-      return (pattern, nil)
+    let colon = self.consume(if: .colon)
+    var lookahead = self.lookahead()
+    var type: RawTypeAnnotationSyntax?
+    if let colon = colon {
+      let result = self.parseResultType()
+      type = RawTypeAnnotationSyntax(
+        colon: colon,
+        type: result,
+        arena: self.arena
+      )
+    } else if allowRecover && lookahead.canParseType() && !self.currentToken.isAtStartOfLine {
```

Comment:
I would swap the conditions because `!self.currentToken.isAtStartOfLine` is by far the cheaper check.

```suggestion
    } else if allowRecover && !self.currentToken.isAtStartOfLine && lookahead.canParseType() {
```